src/Shared/ServerInfrastructure/StringUtilities.cs

```diff
@@ -113,466 +112,429 @@ private static unsafe void GetLatin1StringNonNullCharacters(Span<char> buffer, I
+                output += sizeof(short);
             }
 
-            if (input <= end - sizeof(int))
+            if (input < end)
             {
-                var value = *(int*)input;
-                if (!CheckBytesInAsciiRange(value))
+                if (!TByteCheck.CheckBytes(((sbyte*)input)[0]))
                 {
                     return false;
                 }
-
-                WidenFourAsciiBytesToUtf16AndWriteToBuffer(output, input, value, zero);
-
-                input += sizeof(int);
-                output += sizeof(int);
+                output[0] = (char)input[0];
             }
+
+            return true;
         }
-        else
+
+        Debug.Assert(input + Vector128<sbyte>.Count <= end);
+
+        if (Vector256.IsHardwareAccelerated && input <= end - Vector256<byte>.Count)
         {
-            // 32-bit: Loop ints by default
-            while (input <= end - sizeof(int))
+            var zero256S = Vector256<sbyte>.Zero;
+
+            do
             {
-                var value = *(int*)input;
-                if (!CheckBytesInAsciiRange(value))
+                var vector = Vector256.Load(input);
+                if (!TByteCheck.CheckBytes(vector.AsSByte(), zero256S))
                 {
                     return false;
                 }
 
-                WidenFourAsciiBytesToUtf16AndWriteToBuffer(output, input, value, zero);
+                var (out0, out1) = Vector256.Widen(vector);
+
+                out0.Store((ushort*)output);
+                out1.Store((ushort*)output + Vector256<ushort>.Count);
+
+                input += Vector256<byte>.Count;
+                output += 2 * Vector256<ushort>.Count;
+            } while (input <= end - Vector256<byte>.Count);
 
-                input += sizeof(int);
-                output += sizeof(int);
+            if (input == end)
+            {
+                return true;
             }
         }
 
-        if (input <= end - sizeof(short))
+        var zero128S = Vector128<sbyte>.Zero;
+
+        while (input <= end - Vector128<byte>.Count)
         {
-            if (!CheckBytesInAsciiRange(((short*)input)[0]))
+            var vector = Vector128.Load(input);
+            if (!TByteCheck.CheckBytes(vector.AsSByte(), zero128S))
             {
                 return false;
             }
 
-            output[0] = (char)input[0];
-            output[1] = (char)input[1];
+            var (out0, out1) = Vector128.Widen(vector);
+
+            out0.Store((ushort*)output);
+            out1.Store((ushort*)output + Vector128<ushort>.Count);
 
-            input += sizeof(short);
-            output += sizeof(short);
+            input += Vector128<byte>.Count;
+            output += 2 * Vector128<ushort>.Count;
         }
 
-        if (input < end)
+        if (input != end)
         {
-            if (!CheckBytesInAsciiRange(((sbyte*)input)[0]))
+            // Here we know that from the end at least Vector128<sbyte>.Count elements exists.
```

Comment:
If there are remaining elements, then here we just process it vectorized. This is a simple, but effective heuristic. As there are $[1, 16)$ elements remaining the biggest benefit will be for `15` remaining elements, while for `1` remaining element the benefit won't be that big. Thus one could chose a different threshould $x \in [1, 16)$, but in my tests the overhead introduced by such a calculation is more heavy than just doing it always iff there's a remainder.